src/cpufreq.c

```diff
@@ -72,66 +260,240 @@ static void cpufreq_submit (int cpu_num, double value)
 	sstrncpy (vl.plugin, "cpufreq", sizeof (vl.plugin));
 	sstrncpy (vl.type, "cpufreq", sizeof (vl.type));
 	ssnprintf (vl.type_instance, sizeof (vl.type_instance),
-			"%i", cpu_num);
+			"%zu", cpu_num);
+
+	plugin_dispatch_values (&vl);
+	++reported_last_run;
+}
+
+// used when distribuiton is reported
+static void cpufreq_submit_distribution_value (const char *plugin_instance, long int hertz, derive_t value)
+{
+	value_t values[1];
+	value_list_t vl = VALUE_LIST_INIT;
+
+	values[0].derive = value;
+
+	vl.values = values;
+	vl.values_len = 1;
+	sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+	sstrncpy (vl.plugin, "cpufreq", sizeof (vl.plugin));
+	if (plugin_instance != NULL)
+		sstrncpy (vl.plugin_instance, plugin_instance, sizeof (vl.plugin_instance));
+	sstrncpy (vl.type, "total_time_in_ms", sizeof (vl.type));
+	ssnprintf (vl.type_instance, sizeof (vl.type_instance),
+		   "%ld", hertz);
 
 	plugin_dispatch_values (&vl);
+	++reported_last_run;
 }
 
 static int cpufreq_read (void)
 {
-        int status;
-	unsigned long long val;
 	FILE *fp;
-	char filename[256];
-	char buffer[16];
+	char filename[128];
+	char buffer[512];
+	int status;
+	unsigned long long val;
 
-	for (int i = 0; i < num_cpu; i++)
-	{
-		status = ssnprintf (filename, sizeof (filename),
-				"/sys/devices/system/cpu/cpu%d/cpufreq/"
-				"scaling_cur_freq", i);
-		if ((status < 1) || ((unsigned int)status >= sizeof (filename)))
-			return (-1);
+	reported_last_run = 0;
 
-		if ((fp = fopen (filename, "r")) == NULL)
+	if (!report_distribution) // current value reported only
+	{
+		for (size_t i = 0; i < num_cpu; i++)
 		{
-			char errbuf[1024];
-			WARNING ("cpufreq: fopen (%s): %s", filename,
-					sstrerror (errno, errbuf,
-						sizeof (errbuf)));
-			return (-1);
-		}
+			status = ssnprintf (filename, sizeof (filename),
+					    "/sys/devices/system/cpu/cpu%zu/cpufreq/"
+					    "scaling_cur_freq", i);
+			if ((status < 1) || ((size_t)status >= sizeof (filename)))
+				return (-1);
 
-		if (fgets (buffer, 16, fp) == NULL)
-		{
-			char errbuf[1024];
-			WARNING ("cpufreq: fgets: %s",
-					sstrerror (errno, errbuf,
-						sizeof (errbuf)));
-			fclose (fp);
-			return (-1);
+			if ((fp = fopen (filename, "r")) == NULL)
+			{
+				// cpu could be just switched off
+				// let's check the next one
+				continue;
+			}
+
+			if (fgets (buffer, sizeof(buffer), fp) == NULL)
+			{
+				char errbuf[1024];
+				WARNING ("cpufreq: fgets: %s",
+					 sstrerror (errno, errbuf,
+						    sizeof (errbuf)));
+				fclose (fp);
+				return (-1);
+			}
+
+			if (fclose (fp))
+			{
+				char errbuf[1024];
+				WARNING ("cpufreq: fclose: %s",
+					 sstrerror (errno, errbuf,
+						    sizeof (errbuf)));
+			}
+
+			/* You're seeing correctly: The file is reporting kHz values.. */
+			val = atoll (buffer) * 1000;
+
+			cpufreq_submit_current_value (i, val);
 		}
+	} /* end of if ( !report_distribution ) */
+
+	else // stats on CPU frequency distribution is used
+	{
```

Comment:
This block is quite long and has many levels of indentation. Please move it to its own function.